insights/parsers/initscript.py

```diff
@@ -0,0 +1,104 @@
+"""
+InitScript - files ``/etc/rc.d/init.d/*``
+=========================================
+
+InitScript is a parser for the initscripts in ``/etc/rc.d/init.d``.
+
+Because this parser read multiple files, the initscripts are stored as a list
+within the parser and need to be iterated through in order to find specific
+initscripts.
+
+Examples:
+
+    >>> for initscript in shared[InitScript]: # Parser contains list of all initscripts
+    ...     print "Name:", initscript.file_name
+    ...
+    Name: netconsole
+    Name: rhnsd
+
+"""
+
+from . import SkipException
+from .. import Parser, parser
+import re
+
+CHKCONFIG_REGEX = re.compile(r"^#\s+chkconfig:\s+.*$")
+LSB_REGEX = re.compile(r"^#\s+Provides:\s+.*$")
+
+SHEBANG_REGEX = re.compile(r"^#!\s*/.*$")
+STARTSTOPSTATUS_REGEX = re.compile(r"\b(start|stop|status)\b")
+
+
+class EmptyFileException(SkipException):
+    pass
+
+
+class NotInitscriptException(SkipException):
+    pass
+
+
+@parser("initscript")
+class InitScript(Parser):
+    """
+    Parse initscript files. Each item is a dictionary with following fields:
+
+    Attributes:
+        file_name (str): initscript name
+        file_path (str): initscript path without leading '/'
+        file_content (list): initscript content, line by line
+
+    Because some files may not be real initscripts, to determine whether a file
+    in ``etc/rc.d/init.d/`` is an initscript, the parser checks for
+    ``# chkconfig: <values>`` or ``# Provides: <names>`` strings in the script.
+    If that matches, then it assumes it is an initscript.
+
+    Otherwise, it tries to find out if it is by searching for
+
+    * shebang (e.g. ``#!/bin/bash``) on first line
+    * ``start``/``stop``/``status`` tokens in script
+
+    If 3 or more items are found (half the items searched for + 1), called
+    *confidence* in the code (e.g.  shebang + start + stop), then we assume it
+    is an initscript.
+
+    Otherwise the parser raises a ``SkipException``.
+    """
+
+    def parse_content(self, content):
+        """
+        Raises:
+            EmptyFileException: Raised if file is empty.
+            NotInitscriptException: Raised if likely not an initscript.
+        """
+        self.file_content = content
+
+        # If we find 'chkconfig: XYZ' or 'Provides: NAME', assume we have an
```

Comment:
Hi @rmetrich, how about adjusting the logic a little to save some workloads?
1. Check the `shebang` in content[0] at the very first
2. The check the ``# chkconfig: <values>`` or ``# Provides: <names>`` together with the ``start``/``stop``/``status`` token in the same loop

E.g.

~~~
        # Check for hints (shebang + presence of start/stop/status keywords)
        # Allocate 1 point for each item found. And declare as valid if
        # confidence >= 3 (half of the number of items + 1).
        confidence = 0
        # Check 'shebang' at first
        try:
            line = content[0]
            if SHEBANG_REGEX.match(line):
                confidence += 1
        except IndexError:
            raise EmptyFileException(self.file_path)

        key = {'start': 0, 'stop': 0, 'status': 0}
        for line in content:
            # If we find 'chkconfig: XYZ' or 'Provides: NAME', assume we have an
            # initscript
            if CHKCONFIG_REGEX.match(line) or LSB_REGEX.match(line):
                return
            # Check presence of 'start/stop/status' keywords
            m = STARTSTOPSTATUS_REGEX.search(line)
            if m:
                key[m.group(1)] += 1
        for v in key.itervalues():
            if v != 0:
                confidence += 1

        if confidence < 3:
            raise NotInitscriptException("path: %s, confidence: %d" % (self.file_path, confidence))
~~~

Reply:
Well doing so will cost more.
With the current code, for nicely written initscripts, 5 to 10 lines will be matched against before finding "chkconfig" (regex cost is 2xN, with N ~= 7) 
Whereas with your code proposal, there will be 1 (shebang) + 5xN with N ~= 7

For badly written initscripts, the cost will be 5 x number of lines of the script in all cases.

Reply:
My main target is to merge the two iterators of the content into one loop.
Even the `shebang` check can also be moved into the loop:
~~~
        confidence = 0
        key = {'start': 0, 'stop': 0, 'status': 0}
        for i, line in enumerate(content):
            # If we find 'chkconfig: XYZ' or 'Provides: NAME', assume we have an
            # initscript
            if CHKCONFIG_REGEX.match(line) or LSB_REGEX.match(line):
                return
            if i == 0 and SHEBANG_REGEX.match(line):
                confidence += 1
            # Check presence of 'start/stop/status' keywords
            m = STARTSTOPSTATUS_REGEX.search(line)
            if m:
                key[m.group(1)] += 1
~~~
It basicaly has the same executing order with yours.

Reply:
We can merge the iterations for sure, but it is more CPU costly.

Reply:
:) It's just my suggestion. The logic of your current code works, and it's OK to me that you do not change them.